pymeasure/test.py

```diff
@@ -48,9 +48,13 @@ def expected_protocol(instrument_cls, comm_pairs):
         To represent a response-only communication, use `None` for the command part,
          e.g. `(None, 'RESP1')`.
     """
-    protocol = ProtocolAdapter(comm_pairs)
-    instr = instrument_cls(protocol, "Virtual instrument")
+    instr = instrument_cls("test", name="Virtual instrument",
+                           comm_pairs=comm_pairs)
     yield instr
+    protocol = instr.adapter
+    assert protocol._index == len(comm_pairs), "Not all messages exchanged."
+    assert protocol._write_buffer == b"", "Non empty write buffer."
+    assert protocol._read_buffer == b"", "Non empty read buffer."
```

Comment:
Generally, in exception messages, be helpful and try to give the user information they will need to deal with the exception (often: the content of the variables triggering the exception). It can be quite annoying to get an exception that leaves you only with the knowledge that something weird is going on, and then you have to set a breakpoint there, re-do what triggered the problem, and _then_ inspect the frame contents to find out what's going on. Sometimes, getting the a couple of critical bits of information is enough to unstick you.

If I may suggest something here:
```suggestion
    assert protocol._index == len(comm_pairs), f"Unprocessed protocol definitions remain: {comm_pairs[protocol._index:]}."
    assert protocol._write_buffer == b"", f'Non-empty write buffer: "{protocol._write_buffer}".'
    assert protocol._read_buffer == b"", f'Non-empty read buffer: "{protocol._read_buffer}".'
```
Now the user will immediately see what has been left behind, and maybe already realizes their mistake.

Reply:
Under normal circumstances, I agree. In this case, we're talking about a method used by pytest, which already shows the difference of the failing asserted comparison. For that reason I did not repeat it. On the other hand, it does not hurt.

Reply:
Ah, of course, pytest takes care of that :sweat_smile: